lib/virtio_mmio/virtio_mmio.c

```diff
@@ -29,6 +29,26 @@ extern void virtio_device_reset(struct virtio_device *vdev);
 
 void virtio_mmio_isr(struct virtio_device *vdev);
 
+#if defined(HVL_VIRTIO)
+int virtio_mmio_hvl_wait_cfg(struct virtio_device *vdev, int offset, uint32_t data);
+
+/* weak symbols */
+void virtio_mmio_shm_pool_init(void *mem, size_t size) __attribute__((weak));
+void *virtio_mmio_shm_pool_alloc(size_t size) __attribute__((weak));
+void virtio_mmio_shm_pool_free(void *ptr) __attribute__((weak));
+void virtio_mmio_hvl_ipi(void) __attribute__((weak));
+void virtio_mmio_hvl_wait(uint32_t usec) __attribute__((weak));
+
+/* hypervisorless virtio callback framework */
+static virtio_mmio_hvl_cb_t _cb_funcs[VIRTIO_MMIO_HVL_CB_MAX];
+static void *_cb_args[VIRTIO_MMIO_HVL_CB_MAX];
+static int _num_cb_funcs;
+
+/* pre-shared memory */
+uint32_t __shmem0_pool_start __attribute__((weak));
+uint32_t __shmem0_end __attribute__((weak));
+#endif
```

Comment:
Due to the variable vring size and dynamic placement, the memory pool for shared memory buffers is ultimately defined by __shmem0_pool_start and __shmem0_end. It would be possible to introduce a metal_device_io_region to act as indirection layer, but I don't see any immediate benefit when doing so.